ldms/src/ldmsd/ldmsd_config.c

```diff
@@ -344,24 +413,79 @@ int ldmsd_term_plugin(char *plugin_name)
 	int rc = EINVAL;
 	ldmsd_cfgobj_sampler_t sampler = NULL;
 	ldmsd_cfgobj_store_t store = NULL;
-	ldmsd_plugin_t plug = NULL;
 	ldmsd_cfgobj_t cfgobj;
+	extern struct rbt *cfgobj_trees[];
 
 	sampler = ldmsd_sampler_find(plugin_name);
 	if (sampler) {
-		plug = &sampler->api->base;
+		/*
+		 * Make sure it only has 'find', 'load', 'init', and
+		 * 'cfgobj_tree' references, i.e. it is not running
+		 */
+#ifdef _CFG_REF_DUMP_
+		ref_dump(&sampler->cfg.ref, sampler->cfg.name, stderr);
+#endif
+		ldmsd_cfgobj_lock(&sampler->cfg);
+		if (ldmsd_cfgobj_refcount(&sampler->cfg) > 4) {
+			ovis_log(config_log, OVIS_LERROR,
+				 "The sampler '%s' is in use. Use stop "
+				 "name=%s before attempting to unload "
+				 "the plugin\n",
+				 sampler->cfg.name, sampler->cfg.name);
+			ldmsd_cfgobj_unlock(&sampler->cfg);
+			ldmsd_cfgobj_put(&sampler->cfg, "find");
+			return EBUSY;
+		}
+		/* Deregister all sets for this plugin */
+		struct ldmsd_sampler_set *_set;
+		while (( _set = LIST_FIRST(&sampler->set_list))) {
+			LIST_REMOVE(_set, entry);
+			/* The ldms_set itself will be destroyed by the plugin */
+			free(_set);
+		}
 		cfgobj = &sampler->cfg;
+		ldmsd_sampler_put(sampler, "load");
```

Comment:
Dropping all of these references at once at one level of the code seems to point to something wrong with our code layout. Ideally we would be calling the cleanup functions that pair up with whatever we called to take the refs in the first place. Then the refs can be dropped at the same layer in the code in which they were taken.